### PR TITLE
feat: Add support for structured custom configuration

### PR DESCRIPTION
**Pull request by lenny-intel**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**If your build fails** due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/device-sdk-c/blob/master/.github/CONTRIBUTING.md

## What is the current behavior?
No support for structured custom configuration in TOML and/or Consul

## What is the new behavior?
Now have support for structured custom configuration in TOML and/or Consul
Includes ability to listen for changes on whole or sub-section of the custom configuration when Configuration Provider is enabled.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## New Imports

- [ ] Yes
- [x] No

### CODE REVIEW

File: example/cmd/device-simple/res/configuration.toml [hunk 1]
```diff
@@ -86,3 +86,10 @@ RetryWaitPeriod = "1s"
     Frequency = '30s'
     OnChange = false
     SourceName = 'Image'
+
+# Example structured custom configuration
+[SimpleCustom]
+OnImageLocation = "./res/on.png"
+OffImageLocation = "./res/off.jpg"
+  [SimpleCustom.Writable]
+  DiscoverSleepDurationSecs = 10
```
Comment (tonyespy) [inline example/cmd/device-simple/res/configuration.toml:95]: missing newline @ EOF

File: example/config/configuration.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+package config
+
+import (
+	"errors"
+)
+
+// This file contains example of custom configuration that can be loaded from the service's configuration.toml
+// and/or the Configuration Provider, aka Consul (if enabled).
+// For more details see https://docs.edgexfoundry.org/2.0/microservices/device/Ch-DeviceServices/#custom-configuration
```
Comment (tonyespy) [inline example/config/configuration.go:25]: Is there an existing PR to update the docs and provide the specified page anchor (`#custom-configuration`)?

File: example/driver/simpledriver.go [hunk 3]
```diff
@@ -74,11 +78,55 @@ func (s *SimpleDriver) Initialize(lc logger.LoggingClient, asyncCh chan<- *dsMod
...
+
+	previous := s.serviceConfig.SimpleCustom.Writable
+	s.serviceConfig.SimpleCustom.Writable = *updated
+
+	if reflect.DeepEqual(previous, *updated) {
+		s.lc.Info("No changes detected")
+		return
+	}
+
+	if previous.DiscoverSleepDurationSecs != updated.DiscoverSleepDurationSecs {
```
Comment (tonyespy) [inline example/driver/simpledriver.go:122]: Given that you just did a `DeepEqual` check, I don't think it's possible for these two values to be equal here, so why check? I think your intention was to alert developers that if their `Writable` stanza included more than one field, then using `DeepEqual` makes sense, and there may be cases where some fields remain unchanged? Maybe a comment explaining this would be helpful here?

File: example/driver/simpledriver.go [hunk 4]
```diff
@@ -74,11 +78,55 @@ func (s *SimpleDriver) Initialize(lc logger.LoggingClient, asyncCh chan<- *dsMod
...
+		s.lc.Infof("DiscoverSleepDurationSecs changed to: %d", updated.DiscoverSleepDurationSecs)
+	}
+}
+
 // HandleReadCommands triggers a protocol Read operation for the specified device.
-func (s *SimpleDriver) HandleReadCommands(deviceName string, protocols map[string]contract.ProtocolProperties, reqs []dsModels.CommandRequest) (res []*dsModels.CommandValue, err error) {
+func (s *SimpleDriver) HandleReadCommands(
+	deviceName string, protocols map[string]contract.ProtocolProperties, reqs []dsModels.CommandRequest) (res []*dsModels.CommandValue, err error) {
```
Comment (tonyespy) [inline example/driver/simpledriver.go:129]: Why did you change this? If it's a result of your IDE formatting, please revert and address formatting in a separate commit/PR.

File: pkg/service/service.go [hunk 5]
```diff
@@ -41,21 +44,36 @@ var (
 	ds *DeviceService
 )
 
+// UpdatableConfig interface allows services to have custom configuration populated from configuration stored
+// in the Configuration Provider (aka Consul). Services using custom configuration must implement this interface
+// on their custom configuration, even if they do not use Configuration Provider. If they do not use the
+// Configuration Provider they can have dummy implementation of this interface.
```
Comment (tonyespy) [inline pkg/service/service.go:50]: "dummy implementation" --> "a dummy implementation"

**Review by tonyespy — COMMENTED** [review]
A couple of minor comments inline...

File: example/config/configuration.go (hunk 2 above)
Comment (lenny-intel) [reply example/config/configuration.go:25]: Not yet, but have submitted an issue.

File: example/driver/simpledriver.go (hunk 4 above)
Comment (lenny-intel) [reply example/driver/simpledriver.go]: Not sure. Reverted.

File: example/driver/simpledriver.go (hunk 3 above)
Comment (lenny-intel) [reply example/driver/simpledriver.go:128]: It is just an example of how to check, so if you have multiple setting in the `Writable` section. Comment add

**Review by tonyespy — APPROVED** [review]
LGTM
